klippy/extras/virtual_sdcard.py

```diff
@@ -279,6 +283,12 @@ def work_handler(self, eventtime):
         self.cmd_from_sd = False
         if error_message is not None:
             self.print_stats.note_error(error_message)
+            try:
+                context = self.on_error_gcode.create_template_context()
+                context['error_message'] = error_message
+                self.on_error_gcode.run_gcode_from_command(context)
+            except e:
+                self.print_stats.note_error(str(e))
```

Comment:
Thanks.  However, it's not valid to call `run_gcode_from_command()` outside of a command handler (unless the gcode mutex is obtained and held).

Also, I'm not sure why error_message is passed to the code handler - seems like additional complexity and I'm not sure what anyone will use it for.

Also, I'm not sure why print_stats.note_error() is called on an exception.  Might be better to just call `logging.exception()`.

Also, seems like this config option should have a default - maybe `TURN_OFF_HEATERS` and `M84` (same default as idle_timeout.py).

-Kevin

Reply:
> Thanks. However, it's not valid to call `run_gcode_from_command()` outside of a command handler (unless the gcode mutex is obtained and held).

Apologies Kevin, my grasp for the code is not complete and I forgot about that when I refactored this bit of code. I will fix that ASAP.

> Also, I'm not sure why error_message is passed to the code handler - seems like additional complexity and I'm not sure what anyone will use it for.

My reasoning here was that users might want to manage different errors in different manners.

This way, it would be up to the user if and how he wishes to use `error_message` to do different actions here.

> Also, I'm not sure why print_stats.note_error() is called on an exception. Might be better to just call `logging.exception()`.

I called `print_stats.note_error` so the error would appear on the web client. Without this, any error raised on `on_error_gcode` would not be shown to the user and just appear on the logs instead.

I'm happy to change this if you think it's better just to log the error.

> Also, seems like this config option should have a default - maybe `TURN_OFF_HEATERS` and `M84` (same default as idle_timeout.py).

Fair enough, my though on not putting anything was to keep existing behavior (do nothing!) instead of introducing a breaking change.

If you prefer that way, I will set a default of `TURN_OFF_HEATERS` and `M84` and add that to the configuration changes file so users are aware of this change of behavior.

Reply:
@KevinOConnor Following your comments, I've made some updates, specifically:

- removed the `error_message` parameter so now the template will run on any gcode error raised in virtual_sdcard
- changed from non-mutex guarded `template.run_gcode_from_command()` call to mutex guarded `gcode.run_script(template.render())` call

I also kept the existing "do nothing" default behavior as discussed.